test/spmd/tensor/test_api.py

```diff
@@ -0,0 +1,175 @@
+        )
+        self.assertEqual(sharded_tensor.size(), torch.Size([12, 3]))
+        local_tensor = sharded_tensor.to_local()
+        self.assertEqual(local_tensor.size(), torch.Size([3, 3]))
+
+    @with_comms
+    def test_distribute_tensor_requires_grad(self):
+        device_mesh = DeviceMesh(self.device_type, list(range(self.world_size)))
+        shard_spec = [Shard(0)]
+        tensor_to_shard = torch.randn(12, 3, requires_grad=True)
+        sharded_tensor = distribute_tensor(
+            tensor_to_shard, device_mesh, shard_spec
+        )
+        self.assertTrue(sharded_tensor.requires_grad)
+        self.assertTrue(sharded_tensor.is_leaf)
+        self.assertEqual(sharded_tensor.size(), torch.Size([12, 3]))
+
+    @with_comms
+    def test_distribute_module(self):
+        device_mesh = DeviceMesh(self.device_type, list(range(self.world_size)))
+        # fully shard all linear modules on dim 0
+        module_to_shard = MyModel(20, 20, device=self.device_type)
+        shard_spec = [Shard(0)]
+
+        def shard_fn(name, module):
+            if isinstance(module, nn.Linear):
+                for name, param in module.named_parameters():
+                    dist_param = torch.nn.Parameter(
+                        distribute_tensor(param, device_mesh, shard_spec)
+                    )
+                    module.register_parameter(name, dist_param)
+
+        sharded_module = distribute_module(
+            module_to_shard, device_mesh, shard_fn
+        )
+        for param in sharded_module.parameters():
+            self.assertIsInstance(param, DTensor)
+            self.assertEqual(param.placements, shard_spec)
+
+        replica_spec = [Replicate()]
+        # fully replicate all modules without passing in partition_fn
+        module_to_replicate = MyModel(20, 20, device=self.device_type)
+        replica_module = distribute_module(module_to_replicate, device_mesh)
+        for param in replica_module.parameters():
+            self.assertIsInstance(param, DTensor)
+            self.assertEqual(param.placements, replica_spec)
+
+        # fully replicate all modules by passing in partition_fn
+        def replicate_fn(name, module):
+            if isinstance(module, nn.Linear):
+                for name, param in module.named_parameters():
+                    dist_param = torch.nn.Parameter(
+                        distribute_tensor(param, device_mesh, replica_spec)
+                    )
+                    module.register_parameter(name, dist_param)
+
+        module_to_replicate = MyModel(20, 20, device=self.device_type)
+        replica_module = distribute_module(
+            module_to_replicate, device_mesh, replicate_fn
+        )
+        for param in replica_module.parameters():
+            self.assertIsInstance(param, DTensor)
+            self.assertEqual(param.placements, replica_spec)
+
+        # only shard part of module, and rest of module should be replicate
+        def shard_fn(name, module):
```

Comment:
Ditto, I agree that it's good that we can let users to write whatever shard_fn to have some flexibility. But for cases we know like two-layer linear Megatron style, can we have a more simplified way to shard it with some default util APIs?

Reply:
I think we should provide pre-set shard fns, like you said megatron linear should just be a pre-set function that user can select from, and inside the preset function it could search the model hierarchy and shard the two linears correspondingly.